README.md
# chef-kibana5
[![Cookbook Version](https://img.shields.io/cookbook/v/kibana5.svg)](https://supermarket.chef.io/cookbooks/kibana5)
[![Build Status](https://travis-ci.org/anuriq/chef-kibana5.svg?branch=master)](https://travis-ci.org/anuriq/chef-kibana5)

## Description

This cookbook installs and configures [Kibana](https://www.elastic.co/products/kibana) version 5.2.0 and above. This cookbook may work with Kibana 6.x.x, but no guarantee.

## Requirements

### Cookbooks

- [ark](https://supermarket.chef.io/cookbooks/ark)
- [systemd](https://supermarket.chef.io/cookbooks/systemd)

### Platforms

Tested only on CentOS 7 and Ubuntu 16.04. Should work on any modern Linux OS with systemd.

## Recipes
* `default.rb` - Runs `kibana5_install` and `kibana5_configure` with default options.

## Resources

### kibana5_install

This resource installs Kibana 5 binaries.

#### Attributes

|Attribute|Description|Type|Default|
|---------|-----------|----|-------|
|`name`| Simply name. Currently not used. |String||
|`version`| Version of Kibana. 5.2.0 and above are supported. |String|`5.6.16`|
|`install_method`| Installation method. Currently only `release` is supported. This means installation from official binaries provided by Elastic. |String|`release`|
|`base_dir`| Base directory for Kibana binaries. |String|`/opt/kibana`|
|`svc_user`| Kibana daemon user. |String|`kibana`|
|`svc_group`| Kibana daemon group. |String|`kibana`|

### kibana5_configure

This resource configures Kibana daemon and main configuration file.

#### Attributes

|Attribute|Description|Type|Default|
|---------|-----------|----|-------|
|`svc_name`| Name of the daemon |String||
|`svc_user`| Kibana daemon user. |String|`kibana`|
|`svc_group`| Kibana daemon group. |String|`kibana`|
|`configuration`| Hash with configuration file values. See examples. |Hash||
|`template_cookbook`| Override template from other cookbook. |String|`kibana5`|

examples of configuration hashes:

```ruby
{
  'server.port' => 5601,
  'server.host' => 'localhost',
  'elasticsearch.url' => 'http://localhost:9200',
}
```

```ruby
{
  'server.port' => 5601,
  'server.host' => 'localhost',
  'server.name' => 'My-Kibana',
  'elasticsearch.url' => 'http://localhost:9200',
  'logging.dest' => '/var/log/kibana.log',
  'logging.verbose' => 'true',
  'elasticsearch.requestHeadersWhitelist' => '[ authorization ]',
}
```

## Usage

It is recommended to use resources directly.

## Examples
You may see examples in fixture cookbook:
[test/fixtures/kibana5-test/recipes/default.rb](test/fixtures/kibana5-test/recipes/default.rb)
[test/fixtures/kibana5-test/recipes/nginx.rb](test/fixtures/kibana5-test/recipes/nginx.rb)

## Authors
- Author:: Azat Khadiev (anuriq@gmail.com)


CHANGELOG.md
## 1.6.7 (19 Aug, 2019)
  - Added versions 6.8.1 and 6.8.2 of Kibana;

## 1.6.6 (4 July, 2019)
  - Added version 6.8.0 of Kibana;

## 1.6.5 (27 March, 2019)
  - Add versions 5.6.16 and 6.7.0 of Kibana;
  - Set 5.6.16 as default version of Kibana.
  - Fixed hashes added in v1.6.4

## 1.6.4 (15 March, 2019)
  - Add latest versions of Kibana;

## 1.6.3 (12 October, 2018)
  - Add version 6.4.2 of Kibana;

## 1.6.2 (8 October, 2018)
  - Fix creation of logging folder;
  - Add versions up to 6.4.1 of Kibana;

## 1.6.1 (17 September, 2018)
  - Add versions up to 6.3.0 of Kibana;
  - Fix systemd description property error;
  - Add xpack security flag in Kibana.yml;

## 1.6.0 (4 May, 2018)
  - kibana-plugins are executed from service user
  - allow newest version of `systemd` cookbook
  - fix an issue with helpers for Chef 14
  - rename kitchen.docker.yml to dokken and make it a default testing suite
  - fix issues to make Travis CI builds work

## 1.5.0 (26 Feb, 2018)
  - Fix style according to newest Cookstyle and FoodCritic;
  - Use CentOS 7.4 for testing;
  - Set 5.6.8 as default version of Kibana.

## 1.4.0 (8 Feb, 2018)
  - Add plugin resource;

## 1.3.2 (8 Feb, 2018)
  - Add version 5.6.7 of Kibana;

## 1.3.1 (November 27, 2017)
  - Add option to use custom configuration template from wrapper cookbook;

## 1.3.0 (October 16, 2017)
  - Add version 5.6.2 of Kibana;

## 1.2.0 (August 20, 2017)
  - Add version 5.5.2 of Kibana;
  - Avoid version 3.0 of `systemd` cookbook till it stabilizes.

## 1.1.2 (July 12, 2017)
  - Add reload setting for systemd unit configuration (@kazjote);
  - Fix according to the newest `foodcritic`.

## 1.1.1 (June 20, 2017)
  - Fix use of attributes for default values of resources.

## 1.1.0 (June 8, 2017)
  - Rethink versions and urls placement, move them to helpers instead of attributes;
  - Add some default attributes to simplify configuration control;
  - Remove redundant tests to simplify contributing;
  - Add support for Kibana `5.4.1` and set default version to `5.4.1`.

## 1.0.1 (May 8, 2017)
  - Add support for Kibana `5.4.0`.
  - Set default version of Kibana to `5.4.0`.

## 1.0.0 (April 7, 2017)
  - Set default version of Kibana to `5.3.0`.
  - Change home dir for `kibana` user, this won't run on nodes converged with previous versions, that's why major version is incremented.
  - Add Ubuntu 16.04 for CI testing suites.
  - Configure Travis to ensure CI.

## 0.2.2 (March 9, 2017)
  - Set default version of Kibana to 5.2.2
  - Add some default attributes

## 0.2.1 (February 17, 2017)
  - Set default version of Kibana to 5.2.1

## 0.2.0 (February 14, 2017)
  - Initial cookbook. See README.


test/fixtures/kibana5-test/metadata.rb
name 'kibana5-test'
description 'Testing cookbook kibana5'
version '0.1.0'

depends 'java'
depends 'elasticsearch', '< 4.0'
depends 'nginx', '< 10.0'
depends 'kibana5'


test/integration/nginx/inspec/nginx.rb
title 'Nginx'

describe service('nginx') do
  it { should be_enabled }
  it { should be_running }
end

describe port(80) do
  it { should be_listening }
end

describe command('curl -I http://127.0.0.1/app/kibana') do
  its(:stdout) { should match(%r{HTTP/1.1 200 OK}) }
  its(:stdout) { should match(/kbn-name: kibana/) }
  its(:stdout) { should match(/kbn-version: 5.6.16/) }
end


test/integration/nginx/inspec/kibana.rb
title 'Kibana'

describe user('kibana') do
  it { should exist }
end

describe service('kibana') do
  it { should be_enabled }
  it { should be_running }
end

describe port(5601) do
  it { should be_listening }
end

describe file('/var/log/kibana.log') do
  it { should be_file }
  it { should be_owned_by 'kibana' }
  it { should be_grouped_into 'kibana' }
  its('mode') { should cmp '0644' }
end


resources/install.rb
#
# Cookbook:: kibana5
# Resource:: kibana5_install
#
# Copyright:: 2017, Parallels International GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

resource_name :kibana5_install

property :version, String, default: ''
property :install_method, String, default: 'release'
property :base_dir, String, default: '/opt/kibana'
property :svc_user, String, default: ''
property :svc_group, String, default: ''

default_action :install

action :install do
  version = new_resource.version == '' ? node['kibana5']['version'] : new_resource.version
  svc_user = new_resource.svc_user == '' ? node['kibana5']['service_user'] : new_resource.svc_user
  svc_group = new_resource.svc_group == '' ? node['kibana5']['service_group'] : new_resource.svc_group
  install_method = new_resource.install_method

  distrib_url = kibana_artifact_url(version, install_method)
  distrib_checksum = kibana_artifact_checksum(version, install_method)

  install_dir = ::File.join(new_resource.base_dir, version)

  case install_method
  when 'release'
    group svc_group

    user svc_user do
      comment 'Kibana User'
      gid svc_group
      home new_resource.base_dir
      shell '/bin/bash'
      system true
    end

    directory install_dir do
      mode '0755'
      owner svc_user
      group svc_group
      recursive true
    end

    ark 'kibana' do
      url distrib_url
      version version
      checksum distrib_checksum
      path install_dir
      home_dir ::File.join(install_dir, 'current')
      owner svc_user
    end

    node.default['kibana5']['config_file'] = ::File.join(install_dir, 'current/config/kibana.yml')
    node.default['kibana5']['exec_file'] = ::File.join(install_dir, 'current/bin/kibana')
  else
    raise "Install method #{install_method} is not implemented yet!"
  end
end


resources/plugins.rb
#
# Cookbook:: kibana5
# Resource:: kibana5_plugins
#
# Copyright:: 2017, Parallels International GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

resource_name :kibana5_plugins

property :install_type, String, default: 'native'
property :url, String, default: ''

default_action :install

use_inline_resources if defined?(use_inline_resources)

action :install do
  service 'kibana' do
    action :nothing
  end

  case new_resource.install_type
  when 'native'
    execute "Install #{new_resource.name} Kibana plugin" do
      command "./kibana-plugin install #{new_resource.name}"
      cwd     "/opt/kibana/#{node['kibana5']['version']}/current/bin"
      user    node['kibana5']['service_user']
      notifies :restart, 'service[kibana]'
      not_if "/opt/kibana/#{node['kibana5']['version']}/current/bin/kibana-plugin list | grep #{new_resource.name}"
    end
  when 'url'
    execute "Install #{new_resource.name} Kibana plugin" do
      command "./kibana-plugin install #{new_resource.url}"
      cwd     "/opt/kibana/#{node['kibana5']['version']}/current/bin"
      user    node['kibana5']['service_user']
      notifies :restart, 'service[kibana]'
      not_if "/opt/kibana/#{node['kibana5']['version']}/current/bin/kibana-plugin list | grep #{new_resource.name}"
    end
  else
    Chef::Application.fatal!("Unknown install type: #{new_resource.install_type}")
  end
end


attributes/packages.rb
#
# Cookbook:: kibana5
# Attribute:: packages
#
# Copyright:: 2017, Parallels International GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

default['kibana5']['distribution']['base_url'] = 'https://artifacts.elastic.co/downloads/kibana'

default['kibana5']['distribution']['5.4.1'] = {
  'release' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.4.1-linux-x86_64.tar.gz",
      'checksum' => 'efab73464c6cbc90089acf9275e43c7b3c2bbe8fb62c710567ac523a65c3394b',
    },
    'x86' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.4.1-linux-x86.tar.gz",
      'checksum' => '25b309435ae4760332e2072e80a83848852245f5c91af6f72bf09771a0bd73fa',
    },
  },
  'windows' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.4.1-windows-x86.zip",
      'checksum' => 'bf88a207b5d4be62c02b83cf2d1cce14aac1a48a9feb084817b93e4690b00983',
    },
  },
}

default['kibana5']['distribution']['5.4.0'] = {
  'release' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.4.0-linux-x86_64.tar.gz",
      'checksum' => '58838dcb5b3195455e1eb05926fed27d882a3476370facfc35d39d837940b1bc',
    },
    'x86' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.4.0-linux-x86.tar.gz",
      'checksum' => 'fbeda3f125cd95cc914c1eb49f65313eb55ce65c9300f231f2d050f4c6bbb65f',
    },
  },
  'windows' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.4.0-windows-x86.zip",
      'checksum' => '9c3c36c042962794b1eb9f45a585302f3eb4c01aecca7c06fb53097b7e7d2852',
    },
  },
}

default['kibana5']['distribution']['5.3.0'] = {
  'release' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.3.0-linux-x86_64.tar.gz",
      'checksum' => 'f31369c32d655370f0e51290018004cfbc54e2d648f0f79dba88e36b4d6cce6f',
    },
    'x86' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.3.0-linux-x86.tar.gz",
      'checksum' => 'eddc77334032ed30f98289715781d01e7677d598cf4a3ad5595618cc057a8bfc',
    },
  },
  'windows' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.3.0-windows-x86.zip",
      'checksum' => '56c5fbd5767dea313ab11489cfed6da063677f9ad38fe9c0420535a46bfdbdaf',
    },
  },
}

default['kibana5']['distribution']['5.2.2'] = {
  'release' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.2-linux-x86_64.tar.gz",
      'checksum' => '9c7c526ce286da7f63aed9fb9f9d752ec182ff16bf374c55e55e5a0536a33563',
    },
    'x86' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.2-linux-x86.tar.gz",
      'checksum' => 'de0421728aca00e9d4114b45ad5519b843b700006eb0795fcc2df4371d6baa1d',
    },
  },
  'windows' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.2-windows-x86.zip",
      'checksum' => '8d6435aa6cfef1a8ce8d4a84361139d0ea3848abfe87283765b4da4fbaefdd3f',
    },
  },
}

default['kibana5']['distribution']['5.2.1'] = {
  'release' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.1-linux-x86_64.tar.gz",
      'checksum' => '82f06fc998385f08eac1c26bb0a48b29a34d179fdb7d0e8be216b35782f12bff',
    },
    'x86' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.1-linux-x86.tar.gz",
      'checksum' => 'c3f27b73e29a9b51a89b9b58798a0ea087df29978424255acb441a8b3e3540bf',
    },
  },
  'windows' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.1-windows-x86.zip",
      'checksum' => '8753db954a89a7a5c19a2c93886b523d1f8054bd31ce24c772bd873555a949d8',
    },
  },
}

default['kibana5']['distribution']['5.2.0'] = {
  'release' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.0-linux-x86_64.tar.gz",
      'checksum' => '729af8ab00f719f2038f6c499e508744b274487756e0214b660535ebead6f28a',
    },
    'x86' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.0-linux-x86.tar.gz",
      'checksum' => '9df75d2dc05937064f1ec905edc705ab1a83e8aaddef6b2fa90b2826f6cc2ea4',
    },
  },
  'windows' => {
    'x64' => {
      'url' => "#{node['kibana5']['distribution']['base_url']}/kibana-5.2.0-windows-x86.zip",
      'checksum' => 'bd1be9758b11b071bc033e9ed996598f4da6cdb18b28c475993c429f1c17db31',
    },
  },
}


resources/configure.rb
#
# Cookbook:: kibana5
# Resource:: kibana5_configure
#
# Copyright:: 2017, Parallels International GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

resource_name :kibana5_configure

property :svc_name, String, name_property: true
property :svc_user, String, default: ''
property :svc_group, String, default: ''
property :configuration, Hash, required: true
property :template_cookbook, String, default: 'kibana5'

default_action :configure

action :configure do
  svc_user = new_resource.svc_user == '' ? node['kibana5']['service_user'] : new_resource.svc_user
  svc_group = new_resource.svc_group == '' ? node['kibana5']['service_group'] : new_resource.svc_group

  systemd_service new_resource.svc_name do
    unit_description 'Kibana Backend'
    after %w(network.target remote-fs.target nss-lookup.target)
    install do
      wanted_by 'multi-user.target'
    end
    service do
      environment 'LANG' => 'C'
      user svc_user
      restart 'always'
      exec_start node['kibana5']['exec_file']
      exec_reload '/bin/kill -HUP $MAINPID'
    end
  end

  config = new_resource.configuration

  directory ::File.dirname(config['logging.dest']) do
    owner svc_user
    group svc_group
    mode '0755'
    recursive true
    action :create
    not_if { (config['logging.dest'] == 'stdout') || (config['logging.dest'] == '/var/log/kibana.log') }
  end

  file config['logging.dest'] do
    mode '0644'
    owner svc_user
    group svc_group
    not_if { config['logging.dest'] == 'stdout' }
  end

  template node['kibana5']['config_file'] do
    cookbook new_resource.template_cookbook
    source 'kibana.yml.erb'
    owner svc_user
    group svc_group
    mode '0644'
    variables config: config
    notifies :restart, "service[#{new_resource.svc_name}]"
  end

  service new_resource.svc_name do
    supports start: true, restart: true, stop: true, status: true
    action [:enable, :start]
  end
end


attributes/default.rb
#
# Cookbook:: kibana5
# Attribute:: default
#
# Copyright:: 2017, Parallels International GmbH
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# To override source of Kibana distribution use this, otherwise leave intact
default['kibana5']['url'] = nil
default['kibana5']['checksum'] = nil

default['kibana5']['version'] = '5.6.16'
default['kibana5']['distribution_base_url'] = 'https://artifacts.elastic.co/downloads/kibana'
default['kibana5']['service_user'] = 'kibana'
default['kibana5']['service_group'] = 'kibana'

default['kibana5']['config']['server.port'] = 5601
default['kibana5']['config']['server.host'] = 'localhost'
default['kibana5']['config']['elasticsearch.url'] = 'http://localhost:9200'
default['kibana5']['config']['logging.dest'] = '/var/log/kibana.log'
